src/libmongoc/src/mongoc/mongoc-database.c

```diff
@@ -996,6 +995,293 @@ mongoc_database_create_collection (mongoc_database_t *database,
    return collection;
 }
 
+char *
+_mongoc_get_encryptedField_state_collection (
+   const bson_t *encryptedFields,
+   const char *data_collection,
+   const char *state_collection_suffix,
+   bson_error_t *error)
+{
+   bson_iter_t iter;
+   const char *fieldName = NULL;
+
+   if (0 == strcmp (state_collection_suffix, "esc")) {
+      fieldName = "escCollection";
+   } else if (0 == strcmp (state_collection_suffix, "ecc")) {
+      fieldName = "eccCollection";
+   } else if (0 == strcmp (state_collection_suffix, "ecoc")) {
+      fieldName = "ecocCollection";
+   } else {
+      bson_set_error (error,
+                      MONGOC_ERROR_COMMAND,
+                      MONGOC_ERROR_COMMAND_INVALID_ARG,
+                      "expected state_collection_suffix to be 'esc', 'ecc', or "
+                      "'ecoc', got: %s",
+                      state_collection_suffix);
+      return NULL;
+   }
+
+   if (bson_iter_init_find (&iter, encryptedFields, fieldName)) {
+      if (!BSON_ITER_HOLDS_UTF8 (&iter)) {
+         bson_set_error (error,
+                         MONGOC_ERROR_COMMAND,
+                         MONGOC_ERROR_COMMAND_INVALID_ARG,
+                         "expected encryptedFields.%s to be UTF-8",
+                         fieldName);
+         return NULL;
+      }
+      return bson_strdup (bson_iter_utf8 (&iter, NULL));
+   }
+   return bson_strdup_printf (
+      "enxcol_.%s.%s", data_collection, state_collection_suffix);
+}
+
+static mongoc_collection_t *
+create_collection_with_encryptedFields (mongoc_database_t *database,
+                                        const char *name,
+                                        const bson_t *opts,
+                                        const bson_t *encryptedFields,
+                                        bson_error_t *error)
+{
+   char *escName = NULL;
+   char *eccName = NULL;
+   char *ecocName = NULL;
+   mongoc_collection_t *escCollection = NULL;
+   mongoc_collection_t *eccCollection = NULL;
+   mongoc_collection_t *ecocCollection = NULL;
+   mongoc_collection_t *dataCollection = NULL;
+   bool ok = false;
+   bson_t *cc_opts = NULL;
+   
+   /* Create ESC collection. */
+   escName = _mongoc_get_encryptedField_state_collection (
+      encryptedFields, name, "esc", error);
+   if (!escName) {
+      goto fail;
+   }
+
+   escCollection = create_collection (database, escName, NULL /* opts */, error);
+   if (!escCollection) {
+      goto fail;
+   }
```

Comment:
This pair of repeated ops could be made into a separate function to reduce redundancy and help clarity, as then:

```c
bool okay = 
       (escCollection = _create_encField_state_collection(database, encryptedFields, name, "esc", error)
    && (eccCollection = _create_encField_state_collection(database, encryptedFields, name, "ecc", error)
    && (ecocCollection = _create_encField_state_collection(database, encryptedFields, name, "ecoc", error);
if (!okay) {
    // Failed to create one or more field collections
    goto fail;
}
```